File: sail_server/application/dto/pems.py
# -*- coding: utf-8 -*-
# @file pems.py
# @brief Personal Energy Management System DTOs
# @author sailing-innocent
# @date 2026-03-01
# @version 1.0
# ---------------------------------

"""
个人精力管理系统(PEMS) 数据传输对象
"""

from datetime import date as date_type
from typing import Optional, List, Dict, Any

from pydantic import BaseModel, Field, ConfigDict


class RhythmClass:
    """节律分类常量"""

    WORKDAY = "workday"
    RESTDAY = "restday"
    HOLIDAY = "holiday"
    SICK = "sick"
    TRAVEL = "travel"
    FOCUS = "focus"


class EnergyBudgetResponse(BaseModel):
    """精力预算响应"""

    date: date_type = Field(description="日期")
    day_id: int = Field(description="自然日ID")
    rhythm: str = Field(default="workday", description="节律分类")
    base_energy: int = Field(default=100, description="基础精力")
    health_multiplier: int = Field(default=100, description="健康修正系数(百分比)")
    energy_budget: int = Field(default=100, description="当日精力预算")
    energy_planned: int = Field(default=0, description="已安排任务消耗")
    energy_actual: int = Field(default=0, description="实际已消耗")
    warning_messages: List[str] = Field(default_factory=list, description="预算告警")


class HealthSignalSummary(BaseModel):
    """当日健康信号摘要"""

    sleep_hours: Optional[float] = Field(default=None, description="睡眠时长")
    sleep_quality: Optional[int] = Field(default=None, description="睡眠质量 1-5")
    energy_level: Optional[int] = Field(default=None, description="精力评分 1-5")
    mood: Optional[int] = Field(default=None, description="情绪评分 1-5")
    weight_value: Optional[float] = Field(default=None, description="体重 kg")
    exercise_count: int = Field(default=0, description="运动记录数")


class PemsMissionBrief(BaseModel):
    """PEMS 任务摘要"""

    id: int = Field(description="任务ID")
    name: str = Field(description="任务名称")
    project_id: Optional[int] = Field(default=None, description="项目ID")
    project_name: Optional[str] = Field(default=None, description="项目名称")
    state: int = Field(description="任务状态")
    energy_cost: int = Field(default=0, description="预计精力消耗")
    planned_minutes: int = Field(default=0, description="预计耗时(分钟)")
    health_constraint: str = Field(default="normal", description="健康约束")


class DayViewResponse(BaseModel):
    """某日 PEMS 完整视图"""

    date: date_type = Field(description="日期")
    day_id: int = Field(description="自然日ID")
    rhythm: str = Field(default="workday", description="节律分类")
    energy_budget: EnergyBudgetResponse = Field(description="精力预算")
    health_signals: HealthSignalSummary = Field(
        default_factory=HealthSignalSummary, description="健康信号摘要"
    )
    planned_missions: List[PemsMissionBrief] = Field(default_factory=list, description="已安排任务")
    completed_missions: List[PemsMissionBrief] = Field(default_factory=list, description="已完成任务")
    challenge_checkins: Dict[str, str] = Field(
        default_factory=dict, description="打卡状态"
    )
    insights: List["InsightResponse"] = Field(default_factory=list, description="当日洞察")
    note: Optional[str] = Field(default=None, description="备注")


class TimeSpanViewResponse(BaseModel):
    """周期视图响应"""

    id: int = Field(description="TimeSpan ID")
    class_: str = Field(description="TimeSpan 类型")
    name: str = Field(description="名称")
    start_date: date_type = Field(description="开始日期")
    end_date: date_type = Field(description="结束日期")
    theme: Optional[str] = Field(default=None, description="周期主题")
    energy_capacity: int = Field(default=0, description="总精力预算")
    energy_consumed: int = Field(default=0, description="已消耗精力")
    project_ids: List[int] = Field(default_factory=list, description="关联项目ID")
    health_goals: Dict[str, Any] = Field(default_factory=dict, description="健康目标")
    review_note: Optional[str] = Field(default=None, description="复盘笔记")
    focus_areas: List[str] = Field(default_factory=list, description="关注领域")
    day_count: int = Field(default=0, description="包含天数")


class ProjectTimelineResponse(BaseModel):
    """项目时间线响应"""

    project_id: int = Field(description="项目ID")
    project_name: str = Field(description="项目名称")
    timespan_id: Optional[int] = Field(default=None, description="归属周期ID")
    energy_budget: int = Field(default=0, description="项目精力预算")
    milestones: List[Dict[str, Any]] = Field(default_factory=list, description="里程碑")
    missions: List[PemsMissionBrief] = Field(default_factory=list, description="任务")
    timelogs: List[Dict[str, Any]] = Field(default_factory=list, description="时间投入")


class InsightSeverity:
    """洞察严重级别"""

    INFO = "info"
    WARNING = "warning"
    DANGER = "danger"


class InsightResponse(BaseModel):
    """洞察/建议响应"""

    type: str = Field(description="洞察类型")
    severity: str = Field(default=InsightSeverity.INFO, description="严重级别")
    title: str = Field(description="标题")
    message: str = Field(description="内容")


class PlanMissionRequest(BaseModel):
    """为某日安排任务请求"""

    mission_id: int = Field(description="要安排的任务ID")


class TimeSpanReviewRequest(BaseModel):
    """提交周期复盘请求"""

    review_note: str = Field(description="复盘内容")
    theme: Optional[str] = Field(default=None, description="周期主题")
    focus_areas: Optional[List[str]] = Field(default=None, description="关注领域")


class HealthQuickLogRequest(BaseModel):
    """健康快速录入请求"""

    sleep_hours: Optional[float] = Field(default=None, description="睡眠时长")
    sleep_quality: Optional[int] = Field(default=None, description="睡眠质量 1-5")
    energy_level: Optional[int] = Field(default=None, description="精力评分 1-5")
    mood: Optional[int] = Field(default=None, description="情绪评分 1-5")
    note: Optional[str] = Field(default=None, description="备注")


class StatusResponse(BaseModel):
    """通用状态响应"""

    status: str = Field(default="success", description="状态")
    message: str = Field(default="", description="消息")

File: packages/site/src/lib/data/pems.ts
/**
 * @file pems.ts
 * @brief Personal Energy Management System data types (mapped to Rhythm API)
 * @author sailing-innocent
 * @date 2026-03-01
 */

export interface EnergyBudgetData {
  date: string
  day_id: number
  rhythm: string
  base_energy: number
  health_multiplier: number
  energy_budget: number
  energy_planned: number
  energy_actual: number
  warning_messages: string[]
}

export interface HealthSignalSummaryData {
  sleep_hours: number | null
  sleep_quality: number | null
  energy_level: number | null
  mood: number | null
  weight_value: number | null
  exercise_count: number
}

export interface PemsMissionBriefData {
  id: number
  name: string
  project_id: number | null
  project_name: string | null
  state: number
  energy_cost: number
  planned_minutes: number
  health_constraint: string
}

export interface InsightData {
  type: string
  severity: 'info' | 'warning' | 'danger'
  title: string
  message: string
}

export interface DayViewData {
  date: string
  day_id: number
  rhythm: string
  energy_budget: EnergyBudgetData
  health_signals: HealthSignalSummaryData
  planned_missions: PemsMissionBriefData[]
  completed_missions: PemsMissionBriefData[]
  challenge_checkins: Record<string, string>
  insights: InsightData[]
  note: string | null
}

export interface TimeSpanViewData {
  id: number
  class_: string
  name: string
  start_date: string
  end_date: string
  theme: string | null
  energy_capacity: number
  energy_consumed: number
  project_ids: number[]
  health_goals: Record<string, unknown>
  review_note: string | null
  focus_areas: string[]
  day_count: number
}

export interface ProjectTimelineData {
  project_id: number
  project_name: string
  timespan_id: number | null
  energy_budget: number
  milestones: Array<{
    id: number
    name: string
    date: string | null
    state: number
    energy_weight: number
  }>
  missions: PemsMissionBriefData[]
  timelogs: Array<{
    id: number
    mission_id: number
    day_id: number
    duration_minutes: number
    energy_cost: number
    description: string
  }>
}

export interface HealthQuickLogProps {
  sleep_hours?: number
  sleep_quality?: number
  energy_level?: number
  mood?: number
  note?: string
}

export interface PlanMissionProps {
  mission_id: number
}

export interface TimeSpanReviewProps {
  review_note: string
  theme?: string
  focus_areas?: string[]
}

export const RhythmLabels: Record<string, string> = {
  workday: '工作日',
  restday: '休息日',
  holiday: '假期',
  sick: '病休',
  travel: '出行',
  focus: '专注日',
}

export const RhythmColors: Record<string, string> = {
  workday: 'bg-blue-100 text-blue-700',
  restday: 'bg-green-100 text-green-700',
  holiday: 'bg-yellow-100 text-yellow-700',
  sick: 'bg-red-100 text-red-700',
  travel: 'bg-purple-100 text-purple-700',
  focus: 'bg-indigo-100 text-indigo-700',
}

// ============================================================================
// Raw Rhythm API types (PEMS has been merged into Rhythm)
// ============================================================================

export interface TimeBlockData {
  id: number
  day_id: number
  affair_id: number | null
  block_type: string
  start_time: string
  end_time: string
  status: string
  pinned: boolean
  plan_version: number
  ref: Record<string, unknown>
  affair_title: string | null
  affair_kind: string | null
  energy_cost: number
  ctime?: string
  mtime?: string
}

export interface DomainMinutesData {
  life: number
  work: number
  career: number
}

export interface HealthSignalItemData {
  signal_type: string
  ref_id: number
  value_json: Record<string, unknown>
  htime: string | null
}

export interface CheckinTodayItemData {
  affair: Record<string, unknown>
  done_today: boolean
  last_result: string | null
  week_done_count?: number
  week_target?: number
}

export interface CheckinTodayData {
  date: string
  precepts: CheckinTodayItemData[]
  habits: CheckinTodayItemData[]
}

export interface RhythmDayViewData {
  date: string
  day_id: number
  plan_version: number
  blocks: TimeBlockData[]
  domain_minutes: DomainMinutesData
  energy_consumed: number
  energy_budget: number
  energy_available: number
  buffer_total_minutes: number
  buffer_free_minutes: number
  checkins: CheckinTodayData | null
  health_signals: HealthSignalItemData[]
  insights: string[]
  warnings: string[]
  note: string | null
}

export interface HealthCheckinRequestData {
  collection_type: 'weight' | 'meal' | 'exercise' | 'medication' | 'sleep' | 'mood' | 'energy'
  log_date?: string
  payload: Record<string, unknown>
  note?: string
}

export interface HealthCheckinResponseData {
  id: number
  collection_type: string
  log_date: string
  ref_id: number | null
  affair_id: number | null
  note: string
  created_at?: string
}

export interface EnergyProfileData {
  id: number
  name: string
  daily_energy_budget: number
  curve_template: Record<string, unknown>
  sleep_start: string
  sleep_end: string
  work_hours_cap: number
  spare_time_windows: Record<string, unknown>
  min_buffer_ratio: number
  life_weight: number
  work_weight: number
  career_weight: number
  ctime?: string
  mtime?: string
}

export interface ReviewData {
  id?: number
  scope: string
  period_key: string
  rhythm_score: number
  domain_minutes: Record<string, number>
  precept_compliance_rate: number
  habit_consistency: number
  sleep_window_keeping: number
  venture_budget_fulfillment: number
  buffer_consumed: number
  encroachments: unknown[]
  ai_summary: string
}

// ============================================================================
// Transformers: Rhythm DTO -> legacy PEMS view model (used by site UI)
// ============================================================================

function blockStatusToState(status: string): number {
  if (status === 'done') return 2
  if (status === 'skipped' || status === 'cancelled') return 3
  return 1
}

function deriveRhythm(date: string): string {
  const d = new Date(date)
  const day = d.getDay()
  return day === 0 || day === 6 ? 'restday' : 'workday'
}

export function transformRhythmDayView(raw: RhythmDayViewData): DayViewData {
  const planned_missions: PemsMissionBriefData[] = []
  const completed_missions: PemsMissionBriefData[] = []

  for (const b of raw.blocks) {
    if (!b.affair_id) continue
    const mission: PemsMissionBriefData = {
      id: b.id,
      name: b.affair_title || '未命名',
      project_id: null,
      project_name: null,
      state: blockStatusToState(b.status),
      energy_cost: b.energy_cost,
      planned_minutes: 0,
      health_constraint: '',
    }
    if (b.status === 'done' || b.status === 'skipped') {
      completed_missions.push(mission)
    } else {
      planned_missions.push(mission)
    }
  }

  const health: HealthSignalSummaryData = {
    sleep_hours: null,
    sleep_quality: null,
    energy_level: null,
    mood: null,
    weight_value: null,
    exercise_count: 0,
  }
  for (const s of raw.health_signals) {
    const v = s.value_json
    if (s.signal_type === 'sleep') {
      health.sleep_hours = typeof v.hours === 'number' ? v.hours : null
      health.sleep_quality = typeof v.quality === 'number' ? v.quality : null
    } else if (s.signal_type === 'mood') {
      health.mood = typeof v.score === 'number' ? v.score : null
    } else if (s.signal_type === 'energy') {
      health.energy_level = typeof v.score === 'number' ? v.score : null
    } else if (s.signal_type === 'weight') {
      health.weight_value = typeof v.value_kg === 'number' ? v.value_kg : null
    } else if (s.signal_type === 'exercise') {
      health.exercise_count += 1
    }
  }

  const insights: InsightData[] = raw.insights.map((text) => ({
    type: 'info',
    severity: 'info',
    title: text,
    message: '',
  }))
  raw.warnings.forEach((text) =>
    insights.push({
      type: 'warning',
      severity: 'warning',
      title: text,
      message: '',
    })
  )

  const rhythm = deriveRhythm(raw.date)
  const energyBudget: EnergyBudgetData = {
    date: raw.date,
    day_id: raw.day_id,
    rhythm,
    base_energy: raw.energy_budget,
    health_multiplier: 100,
    energy_budget: raw.energy_budget,
    energy_planned: raw.energy_budget - raw.energy_available,
    energy_actual: raw.energy_consumed,
    warning_messages: raw.warnings,
  }

  return {
    date: raw.date,
    day_id: raw.day_id,
    rhythm,
    energy_budget: energyBudget,
    health_signals: health,
    planned_missions,
    completed_missions,
    challenge_checkins: {},
    insights,
    note: raw.note,
  }
}

export function transformReviewToInsights(review: ReviewData): InsightData[] {
  const insights: InsightData[] = []
  if (review.ai_summary) {
    insights.push({
      type: 'info',
      severity: 'info',
      title: 'AI 复盘',
      message: review.ai_summary,
    })
  }
  if (review.rhythm_score < 60) {
    insights.push({
      type: 'warning',
      severity: 'warning',
      title: '节奏评分偏低',
      message: `当前节奏评分 ${review.rhythm_score.toFixed(0)}，建议回顾今日安排。`,
    })
  }
  if (review.buffer_consumed > 80) {
    insights.push({
      type: 'danger',
      severity: 'danger',
      title: '缓冲过度消耗',
      message: '今日缓冲时间已消耗超过 80%，请保留恢复空间。',
    })
  }
  return insights
}
